BB2-237 Replace asserts with exception handling

dtisza1 opened the pull request:
<!--
You've got a Pull Request you want to submit? Awesome!
This PR template is here to help ensure you're setup for success:
  please fill it out to help ensure that your PR is complete and ready for approval.
-->

**JIRA Ticket:**
[BB2-237](https://jira.cms.gov/browse/BB2-237)


**User Story or Bug Summary:**
<!-- Please copy-paste the brief user story or bug description that this PR is intended to address. -->
As a BB2 Developer, I need to improve exception handling by replacing the usage of ASSERTs in the code base so that I can more easily troubleshoot and alert on edge case issues that arise.


### What Does This PR Do?

<!--
Add detailed description & discussion of changes here.
The contents of this section should be used as your commit message (unless you merge the PR via a merge commit, of course).

Please follow standard Git commit message guidelines:
* First line should be a capitalized, short (50 chars or less) summary.
* The rest of the message should be in standard Markdown format, wrapped to 72 characters.
* Describe your changes in imperative mood, e.g. "make xyzzy do frotz" instead of "[This patch] makes xyzzy do frotz" or "[I] changed xyzzy to do frotz", as if you are giving orders to the codebase to change its behavior.
* List all relevant Jira issue keys, one per line at the end of the message, per: <https://confluence.atlassian.com/jirasoftwarecloud/processing-issues-with-smart-commits-788960027.html>.

Reference: <https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project>.
-->

In general, the ASSERT type statements are converted to exceptions with handling. 

The following is a list of changes from the commits:
  
* Replace assert in apps/capabilities/permissions.py
  - Replace assert with UpstreamServerException in apps/capabilities/permissions.py TokenHasProtectedCapability.has_permission()
  - Create test_protected_path_token_has_no_scope_attribute() unit test to check for exception condition in apps/capabilities/tests.py TestTokenHasProtectedCapabilityScopesSwitchTrue

* Replace asserts in apps/fhir/bluebutton/models.py
  - Replace assert with exception for hash_hicn() and hash_mbi() in in apps/fhir/bluebutton/models.py
  - Add unit tests for test_hash_hicn_empty_string() and test_hash_mbi_empty_string() in apps/fhir/bluebutton/tests/test_models.py
  - Sort imports per convention.

* Replace asserts in apps/mymedicare_cb/models.py
  - Replace assert with exception for get_and_update_user() and create_beneficiary_record() in apps/mymedicare_cb/models.py
  - Change value for missing SLS user_info sub value from None to empty string for authenticate() in apps/mymedicare_cb/views.py
  - Add unit tests for test_callback_exceptions() in apps/mymedicare_cb/tests/test_callback.py. This is to test the new exceptions replacing asserts that can be reproduced via HTTMock calls.
  - Update unit tests for test_fail_create_beneficiary_record() and test_fail_create_multiple_beneficiary_record() in apps/mymedicare_cb/tests/test_models.py to check for new exceptions that replace the previous asserts. Also validate the exception messages.
  - Sort imports per convention.

* Update with custom exception classes
  - Change UpstreamServerException to custom BBCapabilitiesPermissionTokenScopeMissingException class in apps/capabilities/permissions.py and apps/capabilities/tests.py
  - Change UpstreamServerException to custom BBFhirBluebuttonModelException class in apps/fhir/bluebutton/models.py and apps/fhir/bluebutton/tests/test_models.py
  - Change UpstreamServerException to custom BBMyMedicareCallbackCrosswalkCreateException and BBMyMedicareCallbackCrosswalkCreateException classes in apps/mymedicare_cb/models.py and apps/mymedicare_cb/tests/test_models.py
  - Replace hard coded status code numbers in apps/mymedicare_cb/views.py with constants from rest_framework.status to improve readability.
  - Change UpstreamServerException for authenticate() function in apps/mymedicare_cb/views.py to BBMyMedicareCallbackAuthenticateSlsClientException for SLS client connectivity type issues and BBMyMedicareCallbackAuthenticateSlsUserInfoValidateException for SLS user_info validation type issues.  Update tests in apps/mymedicare_cb/tests/test_callback.py
  - Fix bug in apps/mymedicare_cb/views.py authenticate() by moving TRY statement before response_hook_wrapper(). The HTTPError exception was not being caught.


### What Should Reviewers Watch For?

<!--
Add some items to the following list, or remove the entire section if it doesn't apply for some reason.

Common items include:
* Is this likely to address the goals expressed in the user story?
* Are any additional documentation updates needed?
* Are there any unhandled and/or untested edge cases you can think of?
* Is user input properly sanitized & handled?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
-->
***NOTE***: When testing this PR, it might be difficult to reproduce some of the exception conditions. Many can be reproduced by using the MSLS interface locally and/or by manually modifiying `Crosswalk` model entries via the Django ADMIN.  Some conditions may only be reproduced via the unit tests directly calling the get_and_update_user() and create_beneficiary_record() functions.  


If you're reviewing this PR, please check these things, in particular:

* Are there any unhandled and/or untested edge cases you can think of?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
* Naming in general? Do the namings of variables, methods, classes, and other items make sense? Are they too verbose, or not verbose enough? Any redundancies? Are the comments descriptive enough?
* Are audit logging items properly handled? Is sensitive information either not included or hashed?


### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence: N/A.
    * Does this PR add any new software dependencies?  **No**.
    * Does this PR modify or invalidate any of our security controls? **No**.
    * Does this PR store or transmit data that was not stored or transmitted before? **Yes** or **No**.
* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons? **No**.


### What Needs to Be Merged and Deployed Before this PR?

<!--
Add some items to the following list, or remove the entire section if it doesn't apply.

Common items include:
* Database migrations (which should always be deployed by themselves, to reduce risk).
* New features in external dependencies (e.g. BFD).
-->


### Submitter Checklist

<!--
Helpful hint: if needed, Git allows you to edit your PR's commits and history, prior to merge.
See these resources for more information:

* <https://dev.to/maxwell_dev/the-git-rebase-introduction-i-wish-id-had>
* <https://raphaelfabeni.com/git-editing-commits-part-1/>
-->

I have gone through and verified that...:

* [ X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [ X] I have named this PR and its branch such that they'll be automatically be linked to the (most) relevant Jira issue, per: <https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html>.
* [X ] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X ] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X ] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X ] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. BFD, SLS, etc.).
* [X ] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

dtisza1 commented on apps/mymedicare_cb/views.py:
```diff
@@ -93,7 +95,7 @@ def authenticate(request):
 
     # Set identity values from userinfo response.
-    sls_subject = user_info.get("sub", None).strip()
+    sls_subject = user_info.get("sub", "").strip()
```
NOTE:  Was getting an exception here with the strip() function when None. The empty string is good here and will produce an exception down the line.

JFU-GIT commented:
good to see imports convention followed, local tests passed without issues.

njdister commented:
@dtisza1  @JFU-GIT  I'd like to discuss the use of `UpstreamServerException`.  From what I understand, this was primarily used to relay errors that in some way had occurred or would have occurred from our upstream servers (SLS and BFD), however here it looks like its being used a bit more generally.  Is this an off-base thought, or would it be more appropriate to have another exception type?  I just want to make sure we have an easy way to categorize our errors from New Relic.

dtisza1 commented on apps/mymedicare_cb/views.py:
```diff
@@ -72,28 +86,27 @@ def authenticate(request):
-                                    response_hook_wrapper(sender=SLSUserInfoResponse,
-                                                          auth_flow_dict=auth_flow_dict)]})
-
     try:
```
After writing a test for this, found that the HTTPError exception wasn't being caught. So moved the TRY statement up before the requests.get().

dtisza1 commented:
> @dtisza1 @JFU-GIT I'd like to discuss the use of `UpstreamServerException`. From what I understand, this was primarily used to relay errors that in some way had occurred or would have occurred from our upstream servers (SLS and BFD), however here it looks like its being used a bit more generally. Is this an off-base thought, or would it be more appropriate to have another exception type? I just want to make sure we have an easy way to categorize our exceptions from New Relic.

@njdister @JFU-GIT Added more explicit exceptions related to our 11th minute discussion this morning. Also changed some of the result codes to 500 where it made sense, and changed hard coded result codes to use constants from the rest_framework.status.  I really like this!  Do the naming's and usage look OK to you?  TY!

njdister reviewed (approved):
Awesome stuff!  This will help us better categorize and alert on specific types of errors in New Relic.  🌮

JFU-GIT reviewed (approved):
looks good to me, Approved
